### PR TITLE
Don't send reminders for old actions

### PR DESCRIPTION
**Pull request by 1v4n4**
Improvement:
Added 'reminding period'. If for some reason (for example, the service crashed and no one noticed) we have old actions on the queue or in the reminder database, those supporters are not getting reminders anymore.

### CODE REVIEW

File: service/confirm-reminder/src/index.ts [hunk 1]
```diff
@@ -92,24 +98,32 @@ const job = schedule.scheduleJob('* * * * *', async () => {
...
-        const retry = { retry: changeDate(action.action.createdAt, 1, retryArray), attempts: 1 };
-        await db.put<string, RetryRecord>('retry-' + action.actionId, retry, {});
-
-        console.log(`Scheduled confirm reminder: ${action.actionId}`, action);
-      } else {
-        console.error(`Error checking if confirm scheduled in DB`, error);
-        throw error;
+
+    // Don't remind if action from the queue is too old
+    if (retryExpired(action.action.createdAt, maxPeriod)) {
```
Comment (tttp) [inline service/confirm-reminder/src/index.ts:104]: please avoid wrapping everything into ifs, but instead exit early on the opposite, ie:

```
   if (!retryExpired(action.action.createdAt, maxPeriod))
     return;
  // do the normal things if retry isn't expired
```

rationale: 
- avoid deeply nested code and loads of indentation makes it easier to read
- easier to reason, check for errors first and exit early, then the default function path is what it should do by default

File: service/confirm-reminder/src/index.ts [hunk 2]
```diff
@@ -15,12 +17,12 @@ const queueConfirm = process.env.CONFIRM_QUEUE || args.qc || "";
 const queueConfirmed = process.env.CONFIRMED_QUEUE || args.qd || "";
 const emailQueue = process.env.EMAIL_QUEUE || args.qe || "";
 const remindExchange = process.env.REMIND_EXCHANGE || args.qe || "";
-const maxRetries = parseInt(process.env.MAX_RETIRES || args.r || "3");
+const maxRetries = parseInt(process.env.MAX_RETRIES || args.r || "3");
+const maxPeriod = parseInt(process.env.MAX_RETRY_PERIOD || "5");
 const retryArray = (process.env.RETRY_INTERVAL || "2,3").split(",").map(x => parseInt(x)).filter(x => x > 0);
```
Comment (tttp) [inline service/confirm-reminder/src/index.ts:22]: I don't get why maxPeriod is an external param, can't it be generated automatically with 

`const maxPeriod = retryArray.reduce ( (max, d) => (max+d),0)

last but not the least (and not directly linked to this PR:

I don't get the link between maxRetries and retryPeriod. MaxRetries = 3 but only 2 numbers into the retry array. What is happening to the 3rd retry? when is this triggered?

ie. shouldn't maxRetries be retryArray.length?

File: service/confirm-reminder/src/helpers.ts [hunk 3]
```diff
@@ -11,3 +11,15 @@ export const changeDate = (date: string, attempts: number, retryArray : number[]
...
   return new Date(oldDate.setDate(oldDate.getDate() + retryInterval)).toISOString();
 };
+
+export const retryExpired = (date: string, interval: number) => {
+  const today = new Date();
+  today.setDate(today.getDate() - interval);
+
+  const retryDate = new Date(date);
+
+  console.log("aa", date, today, retryDate, interval)
```
Comment (tttp) [inline service/confirm-reminder/src/helpers.ts:21]: let's not leave debug message, or at least with clearer message than "aa"

**Review by tttp — CHANGES_REQUESTED** [review]

File: service/confirm-reminder/src/index.ts (hunk 2 above)
Comment (1v4n4) [reply service/confirm-reminder/src/index.ts:22]: With retries, we are counting all emails sent to supporters, including the first one (when a supporter takes action) which is not exactly a retry.
So, we can use retryArray.length + 1 for maxRetries, if we are sure we will never want to retry more times than that.
Also, we can calculate maxPeriod from retryArray, I see no good reason not to remove two external params.
I'll remove.

File: service/confirm-reminder/src/index.ts (hunk 1 above)
Comment (1v4n4) [reply service/confirm-reminder/src/index.ts]: > please avoid wrapping everything into ifs, but instead exit early on the opposite, ie:
> 
> ```
>    if (!retryExpired(action.action.createdAt, maxPeriod))
>      return;
>   // do the normal things if retry isn't expired
> ```
> 
> rationale:
> 
>     * avoid deeply nested code and loads of indentation makes it easier to read
> 
>     * easier to reason, check for errors first and exit early, then the default function path is what it should do by default

@tttp check if ok now

File: service/confirm-reminder/src/helpers.ts [hunk 4]
```diff
@@ -12,14 +12,12 @@ export const changeDate = (date: string, attempts: number, retryArray : number[]
...
+export const retryValid = (date: string, interval: number) => {
   const today = new Date();
   today.setDate(today.getDate() - interval);
 
   const retryDate = new Date(date);
 
-  console.log("aa", date, today, retryDate, interval)
-
-  return retryDate < today;
+  return retryDate > today;
```
Comment (tttp) [inline service/confirm-reminder/src/helpers.ts:21]: not the opposite, it ">=" is the opposite of "<"

**Review by tttp — APPROVED** [review]

File: service/confirm-reminder/src/helpers.ts (hunk 4 above)
Comment (1v4n4) [reply service/confirm-reminder/src/helpers.ts:21]: Yes! Will fix

**Pull request merged**